doc/modules/grid_search.rst

```diff
@@ -150,6 +154,324 @@ increasing ``n_iter`` will always lead to a finer search.
+
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.model_selection import HalvingGridSearchCV
+    >>> import pandas as pd
+    >>>
+    >>> param_grid = {'max_depth': [3, 5, 10],
+    ...               'min_samples_split': [2, 5, 10]}
+    >>> base_estimator = RandomForestClassifier(random_state=0)
+    >>> X, y = make_classification(n_samples=1000, random_state=0)
+    >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
+    ...                          ratio=2, resource='n_estimators',
+    ...                          max_resources=30, random_state=0).fit(X, y)
+    >>> sh.best_estimator_
+    RandomForestClassifier(max_depth=5, n_estimators=8, random_state=0)
+
+Note that it is not possible to budget on a parameter that is part of the
+parameter grid.
+
+.. _exhausting_the_resources:
+
+Exhausting the available resources
+----------------------------------
+
+As mentioned above, the first iteration uses ``min_resources`` resources. If
+you have a lots of resources available, some of them might be wasted (not
+used)::
+
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.svm import SVC
+    >>> from sklearn.model_selection import HalvingGridSearchCV
+    >>> import pandas as pd
+    >>> param_grid= {'kernel': ('linear', 'rbf'),
+    ...              'C': [1, 10, 100]}
+    >>> base_estimator = SVC(gamma='scale')
+    >>> X, y = make_classification(n_samples=1000)
+    >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
+    ...                          ratio=2).fit(X, y)
+    >>> results = pd.DataFrame(sh.cv_results_)
+    >>> results.groupby('iter')['resource_iter'].unique()
+    iter
+    0    [20]
+    1    [40]
+    2    [80]
+    Name: resource_iter, dtype: object
+
+The search process will only use 80 resources at most, while our maximum
+amount of available resources is ``n_samples=1000``. Note in this case that
+``min_resources = r_0 = 20``. In order for the last iteration to use as many
+resources as possible, you can use the ``force_exhaust_resources``
+parameter.::
+
+    >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
+    ...                            ratio=2, force_exhaust_resources=True,
+    ...                            ).fit(X, y)
+    >>> results = pd.DataFrame.from_dict(sh.cv_results_)
+    >>> results.groupby('iter')['resource_iter'].unique()
+    iter
+    0     [250]
+    1     [500]
+    2    [1000]
+    Name: resource_iter, dtype: object
+
+`min_resources` was here automatically set to 250, which results in the last
+iteration using all the resources. Since ``force_exhaust_resources`` chooses an
+appropriate ``min_resources`` to start with, ``min_resources`` must be set
+to 'auto' (default).
+
+Aggressive elimination of candidates
+------------------------------------
+
+Ideally, we want the last iteration to evaluate ``ratio`` candidates. We then
```

Comment:
It's not clear why we would want to have the latest iteration to evaluate `ratio` candidates.

To me the following schedule could make perfect sense:

iteration | n_samples | n_candidate
-----------|---------------|----------------
1            | 2500          |  40         
2            | 5000          |  20             
3            | 10000        |  10               

Assuming the time complexity of each call to fit is linear in `n_samples`, each iteration of the halving search with ratio=2 would still be balanced and I would better trust the ability of the search to find the best parameters for the full training set out of 10 candidates instead of 2.

Reply:
Shall we introduce an addtitional `min_n_candidates` parameter?

Reply:
We always end up with a single "best" candidate at the end.

The ideal case is when this single candidate is the best out of less than `ratio` candidates, but you still get one left regardless.